test/services/redeem_and_unstake/ProgressService/update.test.ts

```diff
@@ -0,0 +1,329 @@
+    gatewayRepository = sinon.createStubInstance(GatewayRepository, {
+      get: Promise.resolve(gatewayInstance),
+    });
+
+    progressService = new ProgressService(
+      gatewayRepository,
+      originWeb3,
+      auxiliaryWeb3,
+      coGatewayAddress,
+      originWorkerAddress,
+      auxiliaryWorkerAddress,
+    );
+
+    eip20CoGatewayMockObject = {
+      methods: {
+        progressRedeem: () => {},
+      },
+    };
+
+    progressRedeemSpy = sinon.replace(
+      eip20CoGatewayMockObject.methods,
+      'progressRedeem',
+      sinon.fake.returns(progressRedeemRawTx),
+    );
+
+    coGatewaySpy = sinon.replace(
+      interacts,
+      'getEIP20CoGateway',
+      sinon.fake.returns(eip20CoGatewayMockObject),
+    );
+
+    const eip20GatewayMockObject = {
+      methods: {
+        progressUnstake: () => {},
+      },
+    };
+    progressUnstakeSpy = sinon.replace(
+      eip20GatewayMockObject.methods,
+      'progressUnstake',
+      sinon.fake.returns(progressUnstakeRawTx),
+    );
+    gatewaySpy = sinon.replace(
+      interacts,
+      'getEIP20Gateway',
+      sinon.fake.returns(eip20GatewayMockObject as any),
+    );
+
+    utilsSpy = sinon.replace(
+      Utils,
+      'sendTransaction',
+      sinon.fake.resolves(fakeTransactionHash),
+    );
+
+    message = sinon.createStubInstance(Message);
+    message.gatewayAddress = coGatewayAddress;
+    message.messageHash = '0x4223A868';
+    message.secret = '0x123AAF';
+    message.isValidSecret.callsFake(() => true);
+  });
+
+  afterEach(async (): Promise<void> => {
+    sinon.restore();
+  });
+
+  it('should progress on origin and auxiliary when source and target status '
+    + 'is declared', async () => {
+    message.sourceStatus = MessageStatus.Declared;
+    message.targetStatus = MessageStatus.Declared;
+
+    await progressService.update([message]);
+
+    SpyAssert.assert(
+      gatewayRepository.get,
+      1,
+      [[coGatewayAddress]],
+    );
+
+    SpyAssert.assert(
+      utilsSpy,
+      2,
+      [
+        [
+          progressRedeemRawTx,
+          { gasPrice: AUXILIARY_GAS_PRICE, from: auxiliaryWorkerAddress },
+          auxiliaryWeb3,
+        ],
+        [
+          progressUnstakeRawTx,
+          { gasPrice: ORIGIN_GAS_PRICE, from: originWorkerAddress },
+          originWeb3,
+        ],
+      ],
+    );
+
+    SpyAssert.assert(
+      progressRedeemSpy,
+      1,
+      [[message.messageHash, message.secret]],
+    );
+
+    SpyAssert.assert(
+      progressUnstakeSpy,
+      1,
+      [[message.messageHash, message.secret]],
+    );
+
+    SpyAssert.assert(
+      coGatewaySpy,
+      1,
+      [[auxiliaryWeb3, coGatewayAddress]],
+    );
+
+    SpyAssert.assert(
+      gatewaySpy,
+      1,
+      [[originWeb3, gatewayAddress]],
+    );
+
+    sinon.restore();
+  });
+
+  it('should not progress on origin and auxiliary when source status '
+    + 'is not declared', async () => {
+    message.sourceStatus = MessageStatus.Undeclared;
```

Comment:
when we add direction checks in service, we should also add some more negative permutations in test cases

Reply:
Done!